py-polars/polars/functions/eager.py

```diff
@@ -928,29 +931,39 @@ def align_frames(
         raise TypeError(
             "Input frames must be of a consistent type (all LazyFrame or all DataFrame)"
         )
+    on = [on] if (isinstance(on, str) or not isinstance(on, Sequence)) else on
+    align_on = [(c.meta.output_name() if isinstance(c, pl.Expr) else c) for c in on]
 
-    # establish the superset of all "on" column values, sort, and cache
+    # create lazy alignment frame
     eager = isinstance(frames[0], pl.DataFrame)
-    alignment_frame = (
-        concat([df.lazy().select(on) for df in frames])
-        .unique(maintain_order=False)
-        .sort(by=on, descending=descending)
-    )
-    alignment_frame = (
-        alignment_frame.collect().lazy() if eager else alignment_frame.cache()
+    alignment_frame: pl.LazyFrame = (
+        cast(
+            pl.LazyFrame,
+            reduce(
+                lambda x, y: x.lazy().join(
+                    y.lazy(), how="outer", on=align_on, suffix=str(id(y))
+                ),
+                frames,
+            ),
+        )
+        .select(*align_on, F.exclude(align_on))
+        .sort(by=align_on, descending=descending)
     )
-    # finally, align all frames
+
+    # select out aligned frames
+    aligned_cols = set(alignment_frame.columns[len(align_on) :])
+    aligned_frames = []
+    for df in frames:
+        sfx = str(id(df))
+        df_cols = [
+            F.col(f"{c}{sfx}").alias(c) if f"{c}{sfx}" in aligned_cols else F.col(c)
+            for c in df.columns
+        ]
+        aligned_frames.append(alignment_frame.select(*df_cols))
+
     aligned_frames = [
-        alignment_frame.join(
-            other=df.lazy(),
-            on=alignment_frame.columns,
-            how="left",
-        ).select(df.columns)
-        for df in frames
+        df if select is None else df.select(select) for df in aligned_frames
     ]
-    if select is not None:
-        aligned_frames = [df.select(select) for df in aligned_frames]
-
     return [df.collect() for df in aligned_frames] if eager else aligned_frames
```

Comment:
`pl.collect_all` would do things in parallel if I am not mistaken

Reply:
Ahh! I had that in an earlier version and then lost it in a refactor/debugging session... thanks for the catch ;)